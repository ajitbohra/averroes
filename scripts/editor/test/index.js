import "../blocks";
import { parse, serialize } from "../api";

jest.mock(
  "@wordpress/blocks",
  () => ({
    createBlock(name, attributes) {
      return { name, attributes };
    },
    getBlockType(name) {
      return {
        name,
        attributes: {}
      };
    },
    getBlockTypes() {
      [this.getBlockType("core/paragraph")];
    }
  }),
  { virtual: true }
);

const isIdempotent = parsed => {
  expect(parse(serialize(parsed))).toEqual(parsed);
};

describe("integration", () => {
  it("Should parse a paragraph block", () => {
    const markdown = "This is just a simple **markdown** text";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse a paragraph block with extra attributes", () => {
    const markdown = `{% name=core/paragraph {"dropCap": true} %}
This is just a simple ** markdown ** text
{% end %}`;
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse multiple paragraph blocks", () => {
    const markdown = `
		This is just a simple **markdown** text

		This is should produce two paragraphs
`;
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse a heading block", () => {
    const markdown = "## Heading 2nd level";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse multiple headings block", () => {
    const markdown = "## Heading 2nd level\n### Another heading";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse an image block", () => {
    const markdown = "![Image alt text](http://myawesomeimage.tld)";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse an image block with an empty alt", () => {
    const markdown = "![](http://myawesomeimage.tld)";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse an quote block", () => {
    const markdown = "> This is just a simple quote";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse a code block", () => {
    const markdown = "```\nconst thisIsJavaScript = true;\n```";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse a code block using leading spaces", () => {
    const markdown =
      "    const thisIsJavaScript = true;\n    thisIsAlsoCode = true;";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse an image block with extra attributes", () => {
    const markdown = `{% name=core/image {"caption": [], "id": 599} %}
![Textual alternative](http://myawesomeimage.tld)
{% end %}`;
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse an unknown block", () => {
    const markdown =
      '{% name=custom/unknown {"randomAttribute":"randomValue"} %}\n{% end %}';
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });

  it("Should parse a list block", () => {
    const markdown =
      "* This is just a simple list\n* This is another list item";
    const parsed = parse(markdown);

    expect(parsed).toMatchSnapshot();
    isIdempotent(parsed);
  });
});
