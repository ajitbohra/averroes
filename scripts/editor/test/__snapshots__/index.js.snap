// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration Should parse a code block 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": "const thisIsJavaScript = true;",
    },
    "name": "core/code",
  },
]
`;

exports[`integration Should parse a code block using leading spaces 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": "const thisIsJavaScript = true;
thisIsAlsoCode = true;",
    },
    "name": "core/code",
  },
]
`;

exports[`integration Should parse a heading block 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": Array [
        "Heading 2nd level",
      ],
      "level": 2,
    },
    "name": "core/heading",
  },
]
`;

exports[`integration Should parse a list block 1`] = `
Array [
  Object {
    "attributes": Object {
      "values": Array [
        Object {
          "props": Object {
            "children": Array [
              "This is just a simple list",
            ],
          },
          "type": "li",
        },
        Object {
          "props": Object {
            "children": Array [
              "This is another list item",
            ],
          },
          "type": "li",
        },
      ],
    },
    "name": "core/list",
  },
]
`;

exports[`integration Should parse a paragraph block 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": Array [
        "This is just a simple ",
        Object {
          "props": Object {
            "children": Array [
              "markdown",
            ],
          },
          "type": "strong",
        },
        " text",
      ],
    },
    "name": "core/paragraph",
  },
]
`;

exports[`integration Should parse a paragraph block with extra attributes 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": Array [
        "This is just a simple ",
        Object {
          "props": Object {
            "children": Array [
              " markdown ",
            ],
          },
          "type": "strong",
        },
        " text",
      ],
      "dropCap": true,
    },
    "name": "core/paragraph",
  },
]
`;

exports[`integration Should parse an image block 1`] = `
Array [
  Object {
    "attributes": Object {
      "alt": "Image alt text",
      "url": "http://myawesomeimage.tld",
    },
    "name": "core/image",
  },
]
`;

exports[`integration Should parse an image block with an empty alt 1`] = `
Array [
  Object {
    "attributes": Object {
      "alt": "",
      "url": "http://myawesomeimage.tld",
    },
    "name": "core/image",
  },
]
`;

exports[`integration Should parse an image block with extra attributes 1`] = `
Array [
  Object {
    "attributes": Object {
      "alt": "Textual alternative",
      "caption": Array [],
      "id": 599,
      "url": "http://myawesomeimage.tld",
    },
    "name": "core/image",
  },
]
`;

exports[`integration Should parse an quote block 1`] = `
Array [
  Object {
    "attributes": Object {
      "value": Array [
        Object {
          "children": Object {
            "props": Object {
              "children": Array [
                "This is just a simple quote",
              ],
            },
            "type": "p",
          },
        },
      ],
    },
    "name": "core/quote",
  },
]
`;

exports[`integration Should parse an unknown block 1`] = `
Array [
  Object {
    "attributes": Object {
      "randomAttribute": "randomValue",
    },
    "name": "custom/unknown",
  },
]
`;

exports[`integration Should parse multiple headings block 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": Array [
        "Heading 2nd level",
      ],
      "level": 2,
    },
    "name": "core/heading",
  },
  Object {
    "attributes": Object {
      "content": Array [
        "Another heading",
      ],
      "level": 3,
    },
    "name": "core/heading",
  },
]
`;

exports[`integration Should parse multiple paragraph blocks 1`] = `
Array [
  Object {
    "attributes": Object {
      "content": Array [
        "		This is just a simple ",
        Object {
          "props": Object {
            "children": Array [
              "markdown",
            ],
          },
          "type": "strong",
        },
        " text",
      ],
    },
    "name": "core/paragraph",
  },
  Object {
    "attributes": Object {
      "content": Array [
        "		This is should produce two paragraphs",
      ],
    },
    "name": "core/paragraph",
  },
]
`;
